File: guides/micronaut-security-api-key/java/src/main/java/example/micronaut/ApiKeyTokenValidator.java
package example.micronaut;

import io.micronaut.core.async.publisher.Publishers;
import io.micronaut.http.HttpRequest;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.token.validator.TokenValidator;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;

@Singleton // <1>
class ApiKeyTokenValidator implements TokenValidator<HttpRequest<?>>  {

    private final ApiKeyRepository apiKeyRepository;

    ApiKeyTokenValidator(ApiKeyRepository apiKeyRepository) { // <2>
        this.apiKeyRepository = apiKeyRepository;
    }

    @Override
    public Publisher<Authentication> validateToken(String token, HttpRequest<?> request) {
        if (request == null || !request.getPath().startsWith("/api")) { // <3>
            return Publishers.empty();
        }
        return apiKeyRepository.findByApiKey(token)
                .map(principal -> Authentication.build(principal.getName()))
                .map(Publishers::just).orElseGet(Publishers::empty);
    }
}

File: guides/micronaut-security-api-key/python/src/example/micronaut/api_key_token_validator.py
from jakarta.inject import Singleton
from micronaut.core.async_.publisher import Publishers
from micronaut.security.authentication import Authentication
from micronaut.security.token.validator import TokenValidator
from org.reactivestreams import Publisher

from .api_key_repository import ApiKeyRepository


# tag::clazz[]
@Singleton  # <1>
class ApiKeyTokenValidator(TokenValidator):
    def __init__(self, api_key_repository: ApiKeyRepository):  # <2>
        self.api_key_repository = api_key_repository

    def validateToken(self, token: str, request) -> Publisher:
        if request is None or not request.getPath().startswith("/api"):  # <3>
            return Publishers.empty()
        name = self.api_key_repository.find_by_api_key(token)
        if name is None:
            return Publishers.empty()
        return Publishers.just(Authentication.build(name))
# end::clazz[]
